politeiawww/wsdcrdata.go

```diff
@@ -90,9 +100,73 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
 	return nil
 }
 
-// newWSDcrdata return a new wsDcrdata context.
-func newWSDcrdata() (*wsDcrdata, error) {
-	// Init websocket client
+// subToNewBlock subscribes to dcrdata events for new blocks
+func (w *wsDcrdata) subToNewBlock() error {
+	if w.isSubscribed(eventNewBlock) {
+		return errDuplicateSub
+	}
+	_, err := w.client.Subscribe(eventNewBlock)
+	if err != nil {
+		return fmt.Errorf("failed to subscribe: %v", err)
+	}
+	w.addSub(eventNewBlock)
+	log.Debugf("wsDcrdata subscribed to new block")
+	return nil
+}
+
+// reconnect stops the current client, initializes a new one, and subscribes
+// to each of the subscriptions that the previous client was subscribed to.
+func (w *wsDcrdata) reconnect(continueListening chan bool) {
+	previousSubscriptions := make(map[string]struct{})
+	for sub := range w.subscriptions {
+		previousSubscriptions[sub] = struct{}{}
+	}
+
+	w.client.Stop()
+	w.removeAllSubs()
+
+	go func() {
```

Comment:
What happens if the connection goes bad again during this process?

Reply:
A couple points about this function.

1. What happens if the connection goes bad again during this process?
2. In general, you want to minimize nested code whenever possible.  Stuff like not using an `} else {` clause if you don't have to.  It keeps the code cleaner and easier to read.
3. Try to stay away from for loops without an ending condition.  It leads to bugs.

Here's a code snippet with some suggestions.  I haven't tested it at all, but this is the way I would write something like this.

```
func (w *wsDcrdata) reconnect() {
	done := make(chan struct{})

	go func() {
		prevSubscriptions := make(map[string]struct{}, len(w.subscriptions))
		timeToWait := 1 * time.Minute

		for len(prevSubscriptions) > 0 {
			// Reconnect to dcrdata if needed
			err := w.client.Ping()
			if err != nil {
				w.client.Stop()

				// Existing subscriptions have gone bad
				for sub := range w.subscriptions {
					prevSubscriptions[sub] = struct{}{}
				}
				w.removeAllSubs()

				// Reconnect
				c, err := initDcrdataWSClient()
				if err != nil {
					goto wait
				}
				w.client = c
			}

			// Resubscribe
			for sub := range prevSubscriptions {
				if w.isSubscribed(sub) {
					delete(prevSubscriptions, sub)
					continue
				}
				_, err := w.client.Subscribe(sub)
				if err != nil {
					log.Errorf("wsDcrdata failed to subscribe to %v: %v", err)
					goto wait
				}
				w.addSub(sub)
				delete(prevSubscriptions, sub)
			}

		wait:
			log.Debugf("wsDcrdata reconnect waiting %v", timeToWait)
			time.Sleep(timeToWait)
			timeToWait = 2 * timeToWait
		}

		// All previous subscriptions have been re-subscribed
		// to. We're done.
		done <- struct{}{}
	}()

	<-done
}
```